doc/reference/trd-uart.md

```diff
@@ -0,0 +1,523 @@
+is virtualized across multiple clients (e.g., so multiple sources can
+write to the console), individual clients may want to check the baud rate.
+However, they cannot set the baud rate, because that is fixed and shared
+across all of them. Similarly, some services may need to be able to set
+the UART configuration but do not need to check it.
+
+Most devices using serial ports today use 8-bit data, but some older
+devices use more or fewer bits, and hardware implementations support
+this. If the data width of a UART is set to less than 8 bits, data is
+still partitioned into bytes, and the UART sends the least significant
+bits of each byte. Suppose a UART is configured to send 7-bit
+words. If a client sends 5 bytes, the UART will send 35 bits,
+transmitting the bottom 7 bits of each byte. The most significant bit
+of each byte is ignored.
+
+```rust
+pub enum StopBits {
+    One = 1,
+    Two = 2,
+}
+
+pub enum Parity {
+    None = 0,
+    Odd = 1,
+    Even = 2,
+}
+
+pub enum Width {
+    Six = 6,
+    Seven = 7,
+    Eight = 8,
+}
+
+pub struct Parameters {
+    pub baud_rate: u32, // baud rate in bit/s
+    pub width: Width,
+    pub parity: Parity,
+    pub stop_bits: StopBits,
+    pub hw_flow_control: bool,
+}
+
+pub trait Configuration {
+    fn get_baud_rate(&self) -> u32;
+    fn get_width(&self) -> Width;
+    fn get_parity(&self) -> Parity;
+    fn get_stop_bits(&self) -> StopBits;
+    fn get_flow_control(&self) -> bool;
+    fn get_configuration(&self) -> Configuration;
+}
+
+pub trait Configure {
+    fn set_baud_rate(&self, rate: u32) -> Result<u32, ErrorCode>;
+    fn set_width(&self, width: Width) -> Result<(), ErrorCode>;
+    fn set_parity(&self, parity: Parity) -> Result<(), ErrorCode>;
+    fn set_stop_bits(&self, stop: StopBits) -> Result<(), ErrorCode>;
+    fn set_flow_control(&self, on: bool) -> Result<(), ErrorCode>;
+    fn configure(&self, params: Parameters) -> Result<(), ErrorCode>;
+}
+```
+
+Methods in `Configure` can return the following error conditions:
+  - `OFF`: The underlying hardware is currently not available, perhaps
+    because it has not been initialized or in the case of a shared
+    hardware USART controller because it is set up for SPI.
+  - `INVAL`: Baud rate was set to 0.
+  - `ENOSUPPORT`: The underlying UART cannot satisfy this configuration.
+  - `FAIL`: Other failure condition.
+
+
+The UART may be unable to set the precise baud rate specified. For
+example, the UART may be driven off a fixed clock with integer
+prescalar. A call to `configure` MUST set the baud rate to the closest
+possible value to the `baud_rate` field of the `params` argument and a
+call to `set_baud_rate` MUST set the baud rate to the closest possible
+value to the `rate` argument. The `Ok` result of `set_baud_rate`
+includes the actual rate set, while an `Err(INVAL)` result means the
+requested rate is well outside the operating speed of the UART (e.g., 4MHz).
```

Comment:
```suggestion
prescalar. In response to a call to `configure`, the underlying UART MUST set the baud rate to the closest
possible value to that requested by the `baud_rate` field of the `params` argument, and the response to a
call to `set_baud_rate` MUST set the baud rate to the closest possible
value to the provided `rate` argument. The `Ok` result of `set_baud_rate`
includes the actual rate set, while an `Err(INVAL)` result means the
requested rate is well outside the operating speed of the UART (e.g., 4MHz).
```

I eventually got it, but I had to read this many times as my first parse kept interpreting this as the _caller_ need to set the baud rate correctly when calling; this is a bit more verbose but hopefully clearer to other thick-headed readers such as myself 😅

Reply:
Yeah, I get the confusion. But "in response" seems to suggest asynchrony. How about "Calling `configure` MUST ..."

Reply:
Yeah, I get the confusion. But "in response" seems to suggest asynchrony. How about "Calling `configure` MUST ..."